new DNS-SD records

krnowak opened the pull request:
This is a work in progress, not tested or anything, but can serve as a base for API review. It is based on #94, so please ignore the string to number conversion functions. I still have to review this code against the new IETF drafts.

(Edit: Pressed the submit too quickly)

Added three functions:

- generate an ostree avahi service file
  - it follows a bit an API convention in ostree, so it takes a GVariant being a vardict for options
- delete an ostree avahi service file
  - it deletes it for a specific repository index
- delete all ostree avahi service files but the ones in the whitelist
  - this is to be used in the situation when we change the config in a way that we remove some repositories, so we can also remove all the obsolete files without having to remember what files there were.
  - makes me wonder if this is really necessary - probably we could remove the whitelist, so all the ostree avahi service files would be removed and then we could recreate them based on a new configuration

https://phabricator.endlessm.com/T16589

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -110,35 +179,35 @@ generate_avahi_service_template_to_file (GFile *path,
+  g_variant_builder_add (&records_builder, "(sv)",
+                         eos_avahi_v1_head_commit_timestamp,
+                         g_variant_new_string (timestamp_str));
+  txt_records = g_variant_ref_sink (g_variant_builder_end (&records_builder));
```
Would it make more sense to have `generate_avahi_service_template_to_file()` do the `ref_sink` call on `txt_records`?

krnowak replied:
I'd say no, even if it means that we need to worry about sinking the gvariant ourselves and general verbosity of the code. My impression is that functions that actually store the gvariant should call `ref_sink`. In our case the functions simply read contents of the gvariant.

pwithnall replied:
As I understand things, the main thinking behind when to use `ref_sink` is to allow `g_variant_new()` to be used inline where appropriate (which means for most functions which take a `GVariant` as a parameter, like here). For example, I think it would be reasonable to want to call the following (perhaps in a unit test):

`generate_avahi_service_template_to_file (…, g_variant_new_parsed ("@a{sv} { 'blah': … }"), …);`

krnowak replied:
Added the sinking to `generate_avahi_service_template_to_file`.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+  // TODO: Should we store the port number in the configuration
+  // instead having it as the compile-time constant? I can't remember
+  // why I opted for this solution (was it lack of the configuration
+  // file?)
```
Yeah, moving it to a config file would be good. That can be done as a separate commit/PR though, at a fairly low priority.

krnowak replied:
Sure, will keep this TODO there then.

pwithnall replied:
→ Make it a `FIXME`.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+    return FALSE;
+
+  g_hash_table_iter_init (&iter, refs);
+  while (g_hash_table_iter_next (&iter, &key, NULL))
```
Would be nice to give `key` its proper type (`const gchar *`) rather than it remaining a mysterious `gpointer`. That means adding a cast to the `g_hash_table_iter_next()` call though.

krnowak replied:
Alright.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+  service_file = get_ostree_service_file (avahi_service_directory,
+                                          repository_index);
+
+  return generate_ostree_avahi_v1_service_file_from_variants (service_file,
```
You could drop `variant_sink_new()` and have `generate_ostree_avahi_v1_service_file_from_variants()` do the `ref_sink` calls.

krnowak replied:
As stated above - I'm not sure that the function that does not store the gvariant and does not modify it should ever sink or ref it.

krnowak replied:
Added sinking to `variant_to_binary_variant` which gets called by `generate_ostree_avahi_v1_service_file_from_variants` for all the variants it takes.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+                          GCancellable *cancellable,
+                          GError **error)
+{
+  guint32 bloom_size = 255 - strlen(EOS_OSTREE_AVAHI_V1_REFS_BLOOM_FILTER_FIELD) - 2;
```
A comment explaining the maths here would be useful for future reference.

krnowak replied:
Heh, and I got the math wrong here, I think. I need to subtract one byte more (1 for bloom k, one for hash id, one for `=`).

pwithnall replied:
Is this somewhere where a `G_STATIC_ASSERT` to check the total length of something (not necessarily this variable) would be useful to catch mistakes?

/me is a big fan of `G_STATIC_ASSERT`.

krnowak replied:
I like static assert too, but nothing comes to my mind about how we could use it in this case.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+{
+  guint32 bloom_size = 255 - strlen(EOS_OSTREE_AVAHI_V1_REFS_BLOOM_FILTER_FIELD) - 2;
+  guint8 bloom_k = 1;
+  guint8 bloom_hash_id = EOS_OSTREE_AVAHI_STR_BLOOM_HASH_ID;
```
It would probably also be useful to add a comment identifying which parameters need to be kept stable otherwise binary compatibility of the bloom filter between server and client will be broken. (`bloom_size`, `bloom_k` and `bloom_hash_id`.)

krnowak replied:
I'm not sure I understand what you mean here. The default values here can change anytime - the server should handle the change just fine. Or did you mean the `(yyay)` signature?

pwithnall replied:
Sorry, you’re right. I was reading through this too fast. Ignore my comment. :-)

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_SIZE_U,
+                         "u", &bloom_size);
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_K_Y,
+                         "y", &bloom_k);
```
You need to validate that `bloom_k > 0`.

krnowak replied:
Sure.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+  g_autoptr(GVariant) bloom_bits_variant = NULL;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_SIZE_U,
+                         "u", &bloom_size);
```
You need to validate that `bloom_size > 0`.

krnowak replied:
Sure.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+  filter = ostree_bloom_new (bloom_size, bloom_k, bloom_hash_func);
+  if (!fill_filter_with_refs (filter, repo, cancellable, error))
+    return FALSE;
+  sealed_filter = ostree_bloom_seal (filter);
```
OOI, what do you think of the `OstreeBloom` API? The API’s not stable yet, so if you can think of any improvements, please say, and I can get them into the libostree branch before it’s merged.

krnowak replied:
The `ostree_bloom_add_element` function was a bit confusing in the beginning as it was just taking a gpointer and no length. That is - it was until I understood that specifying the hash function for the bloom filter also specifies the element type (in our case - nul-terminated string).

pwithnall replied:
Yup; it’s just the same as `GHashTable` in this respect. I’ll try and make sure that’s covered in the documentation.

krnowak replied:
Maybe consider also adding an API for getting the `k` parameter and a hash function from `OstreeBloom`. I could use it here.

pwithnall replied:
I’ve added `ostree_bloom_get_size()`, `ostree_bloom_get_k()` and `ostree_bloom_get_hash_func()` in the upstream branch.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+    {
+      return FALSE;
+    }
+  g_assert (g_variant_type_equal (g_variant_get_type (refs_bloom_filter_variant),
```
Can simplify this a bit by using `g_variant_is_of_type()`.

krnowak replied:
Yes.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+      break;
+    }
+
+  g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
```
Move this into the `default:` case for symmetry with the other case handling.

krnowak replied:
The other case handling does it in a different way only because there was some more code after `switch`. I can move this inside the default - I hope that the compiler won't start complaining about missing `return`.

pwithnall replied:
The compiler won’t start complaining; it should be able to figure out that there’s a `return` on all code paths from the `switch`.

krnowak replied:
Indeed, it did not complain.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+          gsize idx;
+
+          for (idx = 0; idx < indices_length; ++idx)
+            if (idx == repository_index)
```
This `for; if` construct could just be replaced by `return (repository_index >= indices_length)`, right?

pwithnall replied:
Unless the `idx` is actually supposed to be used as `indices_to_spare[idx]`?

krnowak replied:
Oh, right, thanks for spotting. Yeah, the second comment is right. But all this will be likely gone anyway.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +46,42 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
                                         GCancellable  *cancellable,
                                         GError       **error);
 
+/* new DNS-SD records format for ostree */
```
Eventually it would be good to have some documentation for the format somewhere (in a comment, probably), to save people having to read through the code and work out all the invariants.

krnowak replied:
Will do.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +46,42 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+#define EOS_OSTREE_AVAHI_OPTION_REPO_INDEX_Q "repository-index"
+#define EOS_OSTREE_AVAHI_OPTION_PORT_Q "port"
+
+#define EOS_OSTREE_AVAHI_STR_BLOOM_HASH_ID 1
```
Do all these `#define`s have to be in the header file? They could be private in the C file, right? No reason to de-encapsulate them.

krnowak replied:
Maybe not all the defines, but the `EOS_OSTREE_AVAHI_OPTION` ones could stay here, so the user can use them when specifying options for `eos_ostree_avahi_service_file_generate` instead of writing "repository-index" themselves.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +46,42 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+                                               guint16        repository_index,
+                                               GCancellable  *cancellable,
+                                               GError       **error);
+gboolean eos_ostree_avahi_service_file_cleanup_directory (const gchar   *avahi_service_directory,
```
I think this `cleanup_directory()` API is too complex. There should be no harm in having a `delete_all()` API and then having the caller re-generate the repositories it wants to keep. If that causes unnecessary churn in Avahi’s DNS-SD broadcasts, then I think that’s a bug in Avahi which needs to be fixed (it should probably wait a second after service files are modified before propagating those changes to the network, in case more service file modifications are coming; I don’t know if it does that already).

krnowak replied:
Right, I also had doubts about this API. I will make it simpler.

Also, avahi currently uses inotify (or kqueue on BSDs I think) to watch the services directory. I don't see any timeout here, so I think avahi reloads the config immediately.

pwithnall commented on libeos-updater-util/extensions.h:
```diff
@@ -75,6 +75,7 @@ struct _EosExtensions
 
   GBytes *summary;
   GBytes *summary_sig;
+  guint64 summary_modification_time_unix;
```
I would call this `summary_modification_time_secs` to make it clear what the units are. Maybe add a comment after: `/* since the Unix epoch */`.

krnowak replied:
Ok.

krnowak commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -38,65 +41,131 @@ const gchar * const eos_avahi_v1_ostree_path = "eos_ostree_path";
+  GVariant *txt_value;
+  const gchar *txt_key;
+
+  g_assert (g_variant_type_equal (g_variant_get_type (txt_records),
```
This is a static function, so I think that g_assert is fine. If it fires, then it means I have made a mistake somewhere, as txt records are constructed entirely in this file.

Thanks for the hint about `g_variant_is_of_type` - I can see it is not exactly the same as using `g_variant_type_equal`, but it shouldn't matter in our case.

krnowak commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -38,65 +41,131 @@ const gchar * const eos_avahi_v1_ostree_path = "eos_ostree_path";
+          g_autofree gchar *txt_value_escaped = NULL;
+
+          if (txt_value_data_length + strlen (txt_key) + 1 > 255)
+            g_warning ("The TXT record with key %s is longer than 255 bytes, it will be truncated!",
```
Hmm, bloom size is one of the options we pass to the generator function, so adding an assertion here would mean that the users of the `eos_ostree_avahi_service_file_generate` function will need to do the math themselves (this 255 - 3 - strlen(key)). Unless we implement this `rb0`, `rb1`, `rbN` thing.

pwithnall replied:
The alternative is the callers triggering a `g_warning()` if they don’t do the maths themselves, which is just as bad. Code should be able to work out if it’s going to trigger a warning. I generally fix this by adding a `check` method which just does the validation (and returns a `GError` on failure), which the caller can call before calling this function.

krnowak replied:
I rewrote the code a bit, so now it returns a GError if the record is too long. Also added a `check_options` function that does some basic checks.

krnowak commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -38,65 +41,131 @@ const gchar * const eos_avahi_v1_ostree_path = "eos_ostree_path";
+  //
+  // - more than 256 bytes, because some crappy hardware did not work
+  //   properly with longer records (likely not applicable for us and
+  //   that would limit the size of the bloom filter severely)
```
I don't think so. At least - not on this level, maybe in avahi. But I wrote this after reading [a paragraph in section 6.2 in rfc 6763](https://tools.ietf.org/html/rfc6763#section-6.2).

I believe the symptoms would be simply truncated DNS-SD records sent over the network. I think that we would need to simply state that we don't support crappy hardware (after checking if the hardware we want to run this code on is indeed not crappy ;) )

krnowak replied:
I added another option, so you can specify which size level you don't want to break.

krnowak commented:
Updated, but I did not even try to compile it. Not everything was addressed yet (mostly documenting the txt messages format and the stuff eos extensions file).

krnowak commented:
Now it builds, added a bit of the TXT records format docs. There are some TODOs left, some of which I will address tomorrow. Tests are still to be written.

krnowak commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +46,160 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+gboolean eos_ostree_avahi_service_file_check_options (GVariant  *options,
+                                                      GError   **error);
+gboolean eos_ostree_avahi_service_file_generate (const gchar   *avahi_service_directory,
+                                                 OstreeRepo    *repo,
```
What do you think about replacing the `OstreeRepo *repo` parameter with a string list describing refs and a GDateTime describing repo's summary timestamp? I remember I tried to avoid using ostree API in previous version of DNS-SD records generation API. This would be also tons easier to test. Not sure how to pass them - should them be a part `GVariant *options` (say, "advertised-refs" being an `as` variant and "summary-timestamp" being a `t` variant or rather as explicit parameters to the function. I'm leaning to the latter, as refs and timestamp are not optional and have no sane default values we could use when the user does not specify "advertised-refs" or "summary-timestamp" options.

krnowak replied:
OTOH, passing OstreeRepo allows us to extract some more or just different data from the repo for the future versions of TXT records. Passing refs and summary timestamp explicitly can be risky, because we may possibly not use them in some future version of TXT records.

So maybe just stuffing everything to the options variant is not an entirely bad idea in order to avoid both using ostree API and having useless parameters in future.

pwithnall replied:
I’d go with explicit list-of-refs and timestamp arguments. An options `GVariant` is too flexible (type safety is lost; as you say, both of those parameters are required). Taking an `OstreeRepo` would be slightly nicer, but way too hard to test.

If more data is needed for future versions of the TXT records, we can always add a new API and deprecate this one.

I think when this code is moved into OSTree, we might change the API to take an `OstreeRepo` instead, since it should be easier to test from inside OSTree. But that’s something to worry about later.

krnowak replied:
Ah, I went with the options-only thing. Will rework it later then.

krnowak commented:
Updated to fix two bugs in eos-updater-avahi (I forgot to steal the autoptr variable for out parameter and forgot to terminate the string array with NULL).

pwithnall commented on eos-updater-avahi/docs/eos-updater-avahi.8:
```diff
@@ -117,6 +117,13 @@ Describes the DNS\-SD record for Avahi to advertise to the local network. No
 .\"
+.IP \fI/etc/avahi/services/eos\-ostree\-updater\-[0\-65535].service\fP 4
+.IX Item "/etc/avahi/services/eos-ostree-updater-[0-65535].service"
+Similarly to the eos-updater.service file in the same directory, it
```
`eos-updater.service` needs to be formatted like other filenames in the man page, and the hyphen in it escaped.

krnowak replied:
I changed it to a full path, so I dropped the "in the same directory" part.

pwithnall commented on eos-updater-avahi/docs/eos-updater-avahi.8:
```diff
@@ -117,6 +117,13 @@ Describes the DNS\-SD record for Avahi to advertise to the local network. No
 \fBeos\-updater\-avahi\fP adds, updates and removes when run.
 .\"
+.IP \fI/etc/avahi/services/eos\-ostree\-updater\-[0\-65535].service\fP 4
+.IX Item "/etc/avahi/services/eos-ostree-updater-[0-65535].service"
```
Hyphens need to be escaped.

krnowak replied:
Done.

pwithnall commented on eos-updater-avahi/docs/eos-updater-avahi.8:
```diff
@@ -117,6 +117,13 @@ Describes the DNS\-SD record for Avahi to advertise to the local network. No
+.IP \fI/etc/avahi/services/eos\-ostree\-updater\-[0\-65535].service\fP 4
+.IX Item "/etc/avahi/services/eos-ostree-updater-[0-65535].service"
+Similarly to the eos-updater.service file in the same directory, it
+describes the DNS-SD record for Avahi, but a different format is
```
Hyphen in ‘DNS-SD’ needs to be escaped.

krnowak replied:
Done.

pwithnall commented on eos-updater-avahi/docs/eos-updater-avahi.8:
```diff
@@ -117,6 +117,13 @@ Describes the DNS\-SD record for Avahi to advertise to the local network. No
 advertisement will be made if this file is not present. This is the file which
 \fBeos\-updater\-avahi\fP adds, updates and removes when run.
 .\"
+.IP \fI/etc/avahi/services/eos\-ostree\-updater\-[0\-65535].service\fP 4
```
Use an en-dash in the numeric range: 0–65535. Same in the line below.

krnowak replied:
Done.

pwithnall commented on eos-updater-avahi/eos-updater-avahi.c:
```diff
@@ -23,12 +23,120 @@
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  // TODO: Find the summary without using eos specific bits to ease
```
Make this a `FIXME`, since I’ve been trying to follow the convention in eos-updater that `TODO` is something to fix before a patch is merged, and `FIXME` is something longer-term to fix eventually.

krnowak replied:
Alright.

pwithnall commented on eos-updater-avahi/eos-updater-avahi.c:
```diff
@@ -23,12 +23,120 @@
+                   extensions->summary_modification_time_secs);
+      return FALSE;
+    }
+  // TODO: Try getting the timestamp from summary metadata first.
```
This needs to be fixed. [libostree now adds it to the summary file](https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo.c#L4565)

krnowak replied:
Fixed.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
                                         GError       **error);
 
+/* new DNS-SD records format for ostree */
+/* TODO: gtkdocize it? */
```
Drop this `TODO`. As long as the content’s there, there’s no need to gtkdocize it yet.

krnowak replied:
Ok.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * - summary timestamp,
+ *   - key: "st",
+ *   - type: "t" (uint64)
+ *   - contents: a unix utc timestamp of the summary, ideally telling when
```
Should mention that it’s in seconds (not in milliseconds, for example).

pwithnall replied:
Need to specify its endianness, too.

krnowak replied:
Added.

krnowak replied:
Done. It's big-endian.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * - repository index, a number identifying an OSTree repository
+ *   - key: "ri"
+ *   - type: "q" (uint16)
+ *   - contents: it gets appended to the host's URL as "/%u"; 0 is a special
```
Need to specify its endianness.

krnowak replied:
Alright, it's big endian from now on. I updated the two places where we create gvariants for repo index and summary timestamp. Will file a PR for your ostree branch later with the changes.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * EOS_OSTREE_AVAHI_OPTION_BLOOM_K_Y:
+ *
+ * Specifies the k parameter for the bloom filter. It roughly
+ * translates to how many times an element will be hashed before using
```
Roughly? It means this *exactly*.

krnowak replied:
I wrote roughly because I thought we do something more than just hashing the element `k` times. Fixed.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * Possible values:
+ *
+ * - %EOS_OSTREE_AVAHI_SIZE_LEVEL_CUSTOM
+ * - %EOS_OSTREE_AVAHI_SIZE_LEVEL_SUPPORT_CRAPPY_HARDWARE
```
Maybe s/CRAPPY/FAULTY/

krnowak replied:
Thanks for the suggestion. Applied.

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +44,241 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * that the size limit is specified in the
+ * %EOS_OSTREE_AVAHI_OPTION_TXT_RECORDS_CUSTOM_SIZE_T option.
+ */
+#define EOS_OSTREE_AVAHI_SIZE_LEVEL_CUSTOM ((guint8)0)
```
Why are the `EOS_OSTREE_AVAHI_SIZE_LEVEL_*` values not part of an enum?

krnowak replied:
Uh, I think it was because enums are a bit underspecified (what size are they?), so they may need casting when passed to varargs functions like `g_variant_new`.

pwithnall replied:
IIRC an enum is defined as the smallest signed int type which will contain all the members’ values (I might be wrong here). I agree it’s not immediately obvious though. I’d prefer to put all these values in an enum, cast when passing it to GVariant functions, and add a `G_STATIC_ASSERT (sizeof (TheEnum) <= sizeof (guint8))` or so.

pwithnall replied:
I’m wrong, enums are defined as ints, with some caveats:
http://stackoverflow.com/questions/366017/what-is-the-size-of-an-enum-in-c

pwithnall commented on libeos-updater-util/extensions.c:
```diff
@@ -414,6 +415,22 @@ get_summary (EosExtensions *extensions,
+          return FALSE;
+        }
+    }
+  extensions->summary_modification_time_secs = g_file_info_get_attribute_uint64 (summary_info,
```
It’s possible for `summary_info` to be `NULL` at this point, if the error was `G_IO_ERROR_NOT_FOUND` above.

krnowak replied:
Oops, fixed.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -37,67 +39,355 @@ const gchar * const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+  gsize value_string_length;
+  const gchar *value_string = g_variant_get_string (text_value,
+                                                    &value_string_length);
+  gsize key_length = strlen(key);
```
Add a space before `(`.

krnowak replied:
Fixed.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -37,67 +39,355 @@ const gchar * const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+  gconstpointer value_data = g_variant_get_fixed_array (binary_value,
+                                                        &value_data_length,
+                                                        1);
+  gsize key_length = strlen(key);
```
Missing a space before the `(`.

krnowak replied:
Fixed.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -37,67 +39,355 @@ const gchar * const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_SUPPORT_CRAPPY_HARDWARE:
+      if (total_size > 256)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
```
No point in having separate error strings for all of these cases. You could put the size limits in a `const gsize[]` array indexed by the enum value, then factor out the error handling using a single error string like “TXT records of size %blah break the limit of %blah bytes.”.

krnowak replied:
Fixed. I didn't bother with `const gsize[]` array, because it wouldn't work for custom size, and for lax validation I just do `return TRUE`.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -248,12 +554,697 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+   * for the array of bytes being the bloom filter bits, because it is
+   * the last member of the variant tuple and it is treated specially.
+   */
+  guint32 max_bloom_size = 255 - strlen(EOS_OSTREE_AVAHI_V1_REFS_BLOOM_FILTER_FIELD) - 3;
```
Missing a space before `(`.

krnowak replied:
Fixed.

krnowak commented:
Just pushed the fixes, more may follow up during testing.

krnowak commented:
Pushed some fixes, I still need to make some proper commits out of them. I still did not address the enum-instead-of-defines review issue. At least this code works together with my ostree changes in https://github.com/pwithnall/ostree/pull/2.

pwithnall commented on libeos-update-server/repo.c:
```diff
@@ -814,14 +861,18 @@ handle_path (EusRepo     *self,
-  else if (g_strcmp0 (path, "/config") == 0)
+  else if (g_str_equal (path, "/config"))
     handle_config (self, msg);
+  else if (path_is_summary(path))
```
Missing space before `(`.

krnowak commented:
Updated. I hopefully addressed all the review issues, rebased it on top of current master, updated the ostree bloom implementation to the newest one and prepared the commits.

krnowak commented:
The only issue left is this eos-updater-avahi.py test that fails for me. Will investigate it tomorrow.

pwithnall commented:
This looks good, though given the change in architecture being [discussed on the mailing list](https://mail.gnome.org/archives/ostree-list/2017-May/msg00027.html) we can’t merge this until the new architecture has been finalised, implemented in OSTree, and this PR tweaked to make use of it. :-(

krnowak commented:
I updated it with a most likely broken commit, but the bulk of work should be there. I copied the bloom hash func for OstreeCollectionRef to ostree bloom files. I also copied the ostree-ref.[ch] and modified them slightly to avoid including libglnx and ostree-repo-private.h headers.

mwleeds commented:
I pushed a fixed version of this to the branch "mwleeds/T16589" that compiles and passes the tests. So that's ready for CR.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -248,12 +518,699 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * Create a `.service` file in @avahi_service_directory for the updater. This
+ * instructs Avahi to advertise a DNS-SD service for the updater, with TXT
+ * records indicating this machine has the refs for @ostree_path available with
+ * a commit at @head_commit_timestamp.
```
This documentation seems outdated — `ostree_path` and `head_commit_timestamp` are no longer arguments to this function. We now have `refs_to_advertise`.

pwithnall commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -248,12 +518,699 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ * version of DNS-SD records. For the details, start reading about the
+ * %EOS_OSTREE_AVAHI_OPTION_FORCE_VERSION_Y option. If @options is
+ * %NULL, default values will be used instead. Default values are
+ * described in each options' documentation.
```
s/each options'/each option’s/

pwithnall commented on libeos-updater-util/avahi-service-file.h:
```diff
@@ -44,4 +46,197 @@ gboolean eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+ *
+ * Specifies the k parameter for the bloom filter. It translates to
+ * how many times an element will be hashed before using it to set a
+ * bit in the bloom filter.
```
> It translates to how many times an element will be hashed before using it to set a bit in the bloom filter.

That’s not really correct. The `k` parameter determines how many different hash functions are used. That does mean ‘how many times an element will be hashed’, but the bits which each hash function invocation returns are all set in the bloom filter; not a single bit.

krnowak replied:
https://github.com/endlessm/eos-updater/pull/95#discussion_r116693111 ;)

pwithnall replied:
Heh, snookered.

pwithnall commented:
> I pushed a fixed version of this to the branch "mwleeds/T16589" that compiles and passes the tests. So that's ready for CR.

Since that branch isn’t part of this PR, I can’t easily review it. I think you should be able to force-push to Krzesimir’s branch and push your fixups there so they get included in this PR. Could you try that?

I’ll add comments on specific commits in your branch, which we can hopefully then address in this PR:
 - https://github.com/endlessm/eos-updater/commit/5526c5c06d74f11fa07fa128f72540801e99d182#commitcomment-23636281
 - https://github.com/endlessm/eos-updater/commit/a409c2bd67c5cba76caa22897cdd6f977a2f6d45#commitcomment-23636298
 - https://github.com/endlessm/eos-updater/commit/44895e714465ff1e8abc9bcd04698ee5467b75d2#commitcomment-23636313
 - https://github.com/endlessm/eos-updater/commit/e487d0c2d3d82c7bb37512c0eae367a5e17e7cdd#commitcomment-23636346
 - https://github.com/endlessm/eos-updater/commit/e487d0c2d3d82c7bb37512c0eae367a5e17e7cdd#commitcomment-23636346

krnowak commented:
> Since that branch isn’t part of this PR, I can’t easily review it. I think you should be able to force-push to Krzesimir’s branch and push your fixups there so they get included in this PR. Could you try that?

That won't work unless @mwleeds is the maintainer of the project, no?

mwleeds commented:
@krnowak right I can't push to your `eos-updater` fork, so I opened #101 instead

pwithnall commented:
> @krnowak right I can't push to your eos-updater fork, so I opened #101 instead

Thanks, let’s continue the review there. I’ll close this. Thanks for your updates as well, Krzesimir!

mwleeds commented on libeos-updater-util/avahi-service-file.c:
```diff
@@ -37,67 +39,319 @@ const gchar * const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+  GVariantIter iter;
   g_autoptr(GString) str = NULL;
-  gsize idx;
+  GVariant *txt_value;
```
Similarly for `txt_key`
